lean4-infoview/src/infoview/tooltips.tsx

```diff
@@ -0,0 +1,221 @@
+  // (https://github.com/leanprover/vscode-lean4/issues/51)
+  return ReactDOM.createPortal(popper, document.body)
+})
+
+/** A `<span>` element which gets highlighted when hovered over. It is implemented with JS rather
+ * than CSS in order to allow nesting of these elements. When nested, only the smallest nested
+ * element is highlighted. */
+export const HighlightOnHoverSpan = forwardAndUseRef<HTMLSpanElement, React.HTMLProps<HTMLSpanElement>>((props, ref, setRef) => {
+  const [isPointerOver, setIsPointerOver] = React.useState<boolean>(false)
+  const onPointerOver = (e: React.PointerEvent<HTMLSpanElement>) => {
+    if (ref.current && e.target === ref.current)
+      setIsPointerOver(true)
+  }
+
+  const onPointerOut = (e: React.PointerEvent<HTMLSpanElement>) => {
+    if (ref.current && e.target === ref.current)
+      setIsPointerOver(false)
+  }
+
+  return <span
+      ref={setRef}
+      className={isPointerOver ? 'highlight' : ''}
+      onPointerOver={onPointerOver}
+      onPointerOut={onPointerOut}
+      {...props}
+    >
+      {props.children}
+    </span>
+})
+
+interface TipChainContext {
+  pinParent(): void
+}
+
+const TipChainContext = React.createContext<TipChainContext>({pinParent: () => {}})
+
+/** Shows a tooltip when the children are hovered over or clicked. */
+export const WithTooltipOnHover =
+  forwardAndUseRef<HTMLSpanElement,
+    React.HTMLProps<HTMLSpanElement> & {tooltipContent: TooltipContent}>((props_, ref, setRef) => {
+  const {tooltipContent, ...props} = props_
+
+  // We are pinned when clicked, shown when hovered over, and otherwise hidden.
+  type TooltipState = 'pin' | 'show' | 'hide'
+  const [state, setState] = React.useState<TooltipState>('hide')
+  const shouldShow = state !== 'hide'
+
+  const tipChainCtx = React.useContext(TipChainContext)
+  React.useEffect(() => {
+    if (state === 'pin') tipChainCtx.pinParent()
+  }, [state])
+
+  // Note: because tooltips are attached to `document.body`, they are not descendants of the
+  // hoverable area in the DOM tree, and the `contains` check fails for elements within tooltip
+  // contents. We can use this to distinguish these elements.
+  const isWithinHoverable = (el: EventTarget) => ref.current && ref.current.contains(el as Node)
+  const [logicalDom, logicalDomStorage] = useLogicalDom(ref)
+
+  // We use timeouts for debouncing hover events.
+  const timeout = React.useRef<number>()
+  const clearTimeout = () => {
+    if (timeout.current) {
+      window.clearTimeout(timeout.current)
+      timeout.current = undefined
+    }
+  }
+  const showDelay = 500
+  const hideDelay = 300
+
+  const onClick = (e: React.PointerEvent<HTMLSpanElement>) => {
+    if (!isWithinHoverable(e.target)) return
+    e.stopPropagation()
+    clearTimeout()
+    setState(state => state === 'pin' ? 'hide' : 'pin')
+  }
+
+  React.useEffect(() => {
+    const onClickAnywhere = (e: Event) => {
+      if (!logicalDom.contains(e.target as Node)) {
```

Comment:
```suggestion
      if (e.target instanceof Node && !logicalDom.contains(e.target)) {
```

I'm not 100% sure if this can happen, but it seems good practice regardless.